src/poetry/core/packages/project_package.py

```diff
@@ -3,20 +3,38 @@
 import warnings
 
 from typing import TYPE_CHECKING
-from typing import Any
+from typing import Literal
 from typing import Mapping
 from typing import Sequence
+from typing import TypedDict
 
 from poetry.core.constraints.version import parse_constraint
+from poetry.core.packages.package import Package
+from poetry.core.packages.utils.utils import create_nested_marker
 from poetry.core.version.markers import parse_marker
 
 
 if TYPE_CHECKING:
+    from typing_extensions import NotRequired
+
     from poetry.core.constraints.version import Version
     from poetry.core.packages.dependency import Dependency
 
-from poetry.core.packages.package import Package
-from poetry.core.packages.utils.utils import create_nested_marker
+    SupportedPackageFormats = Literal["sdist", "wheel"]
+
+    BuildConfigSpec = TypedDict(
+        "BuildConfigSpec",
+        {"script": NotRequired[str], "generate-setup-file": NotRequired[bool]},
+    )
+
+    class PackageSpec(TypedDict):
+        include: str
+        to: str
+        format: list[SupportedPackageFormats]
+
+    class IncludeSpec(TypedDict):
+        path: str
+        format: list[SupportedPackageFormats]
```

Comment:
Since these were `Mappings` before and thus immutable for a reason (see comment in `ProjectPackage.__init__`), we probably should declare all attributes as [ReadOnly](https://typing.readthedocs.io/en/latest/spec/typeddict.html#typing-readonly-type-qualifier).